File: solutions/C++/design_search_words_iterative.cpp
#include <iostream>
#include <unordered_map>
#include <queue>
#include <string>

struct TrieNode {
    std::unordered_map<char, TrieNode*> children;
    bool is_end_of_word = false;
};

class WordDictionary {
private:
    TrieNode* root;

public:
    WordDictionary() {
        root = new TrieNode();
    }

    void addWord(const std::string& word) {
        TrieNode* node = root;
        for (char ch : word) {
            if (node->children.find(ch) == node->children.end()) {
                node->children[ch] = new TrieNode();
            }
            node = node->children[ch];
        }
        node->is_end_of_word = true;
    }

    bool search(const std::string& word) {
        std::queue<std::pair<TrieNode*, int>> q;
        q.push({root, 0});

        while (!q.empty()) {
            auto [node, index] = q.front();
            q.pop();

            if (index == (int)word.length()) {
                if (node->is_end_of_word) {
                    return true;
                }
                continue;
            }

            char ch = word[index];
            if (ch == '.') {
                for (auto& [c, child] : node->children) {
                    q.push({child, index + 1});
                }
            } else {
                if (node->children.find(ch) != node->children.end()) {
                    q.push({node->children[ch], index + 1});
                }
            }
        }
        return false;
    }
};

int main() {
    WordDictionary wd;
    wd.addWord("bad");
    wd.addWord("dad");
    wd.addWord("mad");
    std::cout << wd.search("pad") << std::endl;  // 0
    std::cout << wd.search("bad") << std::endl;  // 1
    std::cout << wd.search(".ad") << std::endl;  // 1
    std::cout << wd.search("b..") << std::endl;  // 1
    return 0;
}

File: solutions/Golang/design_search_words_iterative.go
package main

import "fmt"

type TrieNode struct {
	children    map[rune]*TrieNode
	isEndOfWord bool
}

type WordDictionary struct {
	root *TrieNode
}

func NewWordDictionary() *WordDictionary {
	return &WordDictionary{
		root: &TrieNode{
			children: make(map[rune]*TrieNode),
		},
	}
}

func (wd *WordDictionary) AddWord(word string) {
	node := wd.root
	for _, ch := range word {
		if _, exists := node.children[ch]; !exists {
			node.children[ch] = &TrieNode{
				children: make(map[rune]*TrieNode),
			}
		}
		node = node.children[ch]
	}
	node.isEndOfWord = true
}

func (wd *WordDictionary) Search(word string) bool {
	type state struct {
		node  *TrieNode
		index int
	}

	queue := []state{{wd.root, 0}}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current.index == len(word) {
			if current.node.isEndOfWord {
				return true
			}
			continue
		}

		ch := rune(word[current.index])
		if ch == '.' {
			for _, child := range current.node.children {
				queue = append(queue, state{child, current.index + 1})
			}
		} else {
			if child, exists := current.node.children[ch]; exists {
				queue = append(queue, state{child, current.index + 1})
			}
		}
	}
	return false
}

func main() {
	wd := NewWordDictionary()
	wd.AddWord("bad")
	wd.AddWord("dad")
	wd.AddWord("mad")
	fmt.Println(wd.Search("pad"))  // false
	fmt.Println(wd.Search("bad"))  // true
	fmt.Println(wd.Search(".ad"))  // true
	fmt.Println(wd.Search("b.."))  // true
}
